src/shell/history.rs

```diff
@@ -58,4 +102,89 @@ impl<'a> ShellHistory for Shell<'a> {
             context.history.set_file_name(None);
         }
     }
+
+    fn save_command_in_history(&mut self, command: &str) {
+        if self.should_save_command(command) {
+            // Mark the command in the context history
+            self.set_context_history_from_vars();
+            if let Err(err) = self.context.as_mut().unwrap().history.push(command.into()) {
+                let stderr = io::stderr();
+                let mut stderr = stderr.lock();
+                let _ = writeln!(stderr, "ion: {}", err);
+            }
+        }
+    }
+
+    fn update_ignore_patterns(&mut self, patterns: &Array) {
+        let mut flags = IgnoreFlags::empty();
+        let mut regexes = Vec::new();
+        // for convenience and to avoid typos
+        let regex_prefix = "regex:";
+        for pattern in patterns {
+            match pattern.as_ref() {
+                "all" => flags |= IGNORE_ALL,
+                "no_such_command" => flags |= IGNORE_NO_SUCH_COMMAND,
+                "whitespace" => flags |= IGNORE_WHITESPACE,
+                // The length check is there to just ignore empty regex definitions
+                _ if pattern.starts_with(regex_prefix) && pattern.len() > regex_prefix.len() => {
+                    flags |= IGNORE_BASED_ON_REGEX;
+                    let regex_string = pattern[regex_prefix.len()..].to_owned();
+                    // We save the compiled regexes, as compiling them can be  an expensive task
+                    if let Ok(regex) = Regex::new(&regex_string) {
+                        regexes.push(regex);
+                    }
+                }
+                _ => continue,
+            }
+        }
+
+        self.ignore_setting.flags = flags;
+        self.ignore_setting.regexes = if regexes.len() > 0 {
+            Some(regexes)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> ShellHistoryPrivate for Shell<'a> {
+    fn should_save_command(&self, command: &str) -> bool {
+        // just for convenience and to make the code look a bit cleaner
+        let ignore = &self.ignore_setting.flags;
+        let regexes = &self.ignore_setting.regexes;
+
+        // without the second check the command which sets the environment variable would also be
+        // ignored. However, this behavior might not be wanted.
+        if ignore.contains(IGNORE_ALL) && !command.contains("HISTORY_IGNORE") {
+            return false;
+        }
+
+        // Here we allow to also ignore the setting of the environment variable because we assume
+        // the user entered the leading whitespace on purpose.
+        if ignore.contains(IGNORE_WHITESPACE) {
+            if let Some(c) = command.chars().next() {
+                if c.is_whitespace() {
+                    return false;
+                }
+            }
```

Comment:
Can rewrite this block as:

```rust
if command.bytes().next().map_or(false, |b| (b as char).is_whitespace()) {
    return true;
}
```

Reply:
Shouldn't we use `command.chars()` here?

We wouldn't need the cast and nothing breaks if the first character is some multi-byte unicode char.

Reply:
Has higher performance to use bytes. Whitespace characters are never larger than a byte, so it's a bit more efficient.